src/models/predict_large_image.py

```diff
@@ -0,0 +1,59 @@
+import argparse
+from deepforest import get_data
+from deepforest import preprocess
+
+import torch
+from torchvision.ops import nms
+import cv2 as cv
+import numpy as np
+import pandas as pd
+
+from utils.plotting import plot_bboxes
+from models.predict_model import get_model
+
+
+def predict_large_image(
+    numpy_image, model, patch_size=400, patch_overlap=0.05, iou_threshold=0.1
+):
+    # crop original image
+    windows = preprocess.compute_windows(
+        numpy_image=numpy_image, patch_size=patch_size, patch_overlap=patch_overlap
+    )
+
+    predictions_list = []
+
+    for window in windows:
+        crop = numpy_image[window.indices()]
+        pred_df = model.predict_image(crop.astype(np.float32))
+
+        # adjust crop offset relative to the original image
+        pred_df["xmin"] = window.x + pred_df["xmin"]
+        pred_df["ymin"] = window.y + pred_df["ymin"]
+        pred_df["xmax"] = (window.x + window.w) - (patch_size - pred_df["xmax"])
+        pred_df["ymax"] = (window.y + window.h) - (patch_size - pred_df["ymax"])
+
+        predictions_list.append(pred_df)
+
+    final_preds_df = pd.concat(predictions_list)
+    final_preds_df.reset_index(drop=True, inplace=True)
+
+    # filter out duplicate bboxes as a result of crops overlap by applying NMS
+    bboxes = final_preds_df[["xmin", "ymin", "xmax", "ymax"]].values
+    confs = final_preds_df["score"].values
+
+    ixs = nms(torch.tensor(bboxes), torch.tensor(confs), iou_threshold=iou_threshold)
```

Comment:
this assumes confidence would decrease if lets say only a half of the tree is seen from a photo. Which i don't think that it is reliable to assume.
So here, we keep higher confidence predictions, but i believe that the confidence should also be scaled by the closeness to the center of the window.
e.g.
middle of window, before nms, confidence = confidence * 1.0
margin of window, before nms, confidence = confidence * 0.5 (or other smaller scale)
This would guide NMS to keep detections which are closer to the center of the window, on the idea that the closer to the source window's center, it's more probable that the whole tree is visible.

What do you think?